rpc: allow specifying min chain depth for inputs in fund calls

champo opened the pull request:
Enables users to craft BIP-125 replacements with changes to the output
list, ensuring that if additional funds are inputs they will be added.

champo committed:
rpc: allow specifying min chain depth for inputs in fund calls

Enables users to craft BIP-125 replacements with changes to the output
list, ensuring that if additional funds are needed they will be added.

achow101 commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -969,6 +970,53 @@ def test_include_unsafe(self):
+        utxo1 = self.nodes[0].listunspent(0, 1, [unconfirmedAddress])[0]
+        assert unconfirmedTxid == utxo1['txid']
+
+        self.log.info("Crafting PSBT using an unconfirmed input")
```
In 6086c58d27146d500b7ff8de89a5279a6f88457d "rpc: allow specifying min chain depth for inputs in fund calls"

```suggestion
        self.log.info("Crafting tx using an unconfirmed input")
```

Here and elsewhere in this file too.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3215,6 +3224,7 @@ static RPCHelpMan fundrawtransaction()
                             {"include_unsafe", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include inputs that are not safe to spend (unconfirmed transactions from outside keys and unconfirmed replacement transactions).\n"
                                                           "Warning: the resulting transaction may become invalid if one of the unsafe inputs disappears.\n"
                                                           "If that happens, you will need to fund the transaction with different inputs and republish it."},
+                            {"inputs_min_depth", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this chain depth."},
```
IMO should be `minconf`, see `listunspent` RPC.

achow101 reviewed (commented):
Seems reasonable, although the use case for this is not immediately obvious to me.

Perhaps it would also be useful to add `inputs_max_depth` too?

promag reviewed (commented):
Concept ACK.

champo commented:
> Seems reasonable, although the use case for this is not immediately obvious to me.

The use case is for applications that are using RBF for something else than bumping fee (as implemented by `bumpfee` RPC). RBF requires new inputs added to a replacement to be confirmed. The functional tests are modeled after one of the use cases: increasing the value of an output beyond the current funding of the TX.

> Perhaps it would also be useful to add `inputs_max_depth` too?

Will do!

champo committed:
test: fix inputs_min_depth test cases after rebase

champo committed:
rpc: rename inputs_min_depth to minconfs for consistency with listunspent

champo committed:
rpc: add maxconfs parameter in fund calls

champo commented:
Thanks for the quick look guys! I've updated the changset with your feedback. I'm not sure what the ettiquete is here regarding squashing mid-review, so I opted to add new commits. LMK if this is not what you expect.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4043,6 +4063,8 @@ static RPCHelpMan send()
                     {"include_unsafe", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include inputs that are not safe to spend (unconfirmed transactions from outside keys and unconfirmed replacement transactions).\n"
                                                           "Warning: the resulting transaction may become invalid if one of the unsafe inputs disappears.\n"
                                                           "If that happens, you will need to fund the transaction with different inputs and republish it."},
+                    {"minconfs", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this confirmations."},
```
It's singular:
https://github.com/bitcoin/bitcoin/blob/b295395664bd37e26d168c329f238237b34aef8c/src/wallet/rpcwallet.cpp#L2844-L2845

champo committed:
scripted-diff: drop trailing s from maxconfs and minconfs

-BEGIN VERIFY SCRIPT-
sed -i"" -e "s/minconfs/minconf/" ./src/wallet/rpcwallet.cpp ./test/functional/rpc_fundrawtransaction.py ./test/functional/rpc_psbt.py
sed -i"" -e "s/maxconfs/maxconf/" ./src/wallet/rpcwallet.cpp ./test/functional/rpc_fundrawtransaction.py ./test/functional/rpc_psbt.py
-END VERIFY SCRIPT-

luke-jr commented:
Duplicate of #14641 ?

luke-jr commented:
Rebased & squashed here: https://github.com/bitcoin/bitcoin/compare/master...luke-jr:rpc_fundtx_minmaxconf

fanquake commented:
@champo would you like to follow up here?

glozow commented:
Closing as this has needed rebase for more than 1 year. Feel free to reopen if you get a chance to work on this again in the future, thanks!
